### PR TITLE
[Request For Comments] Adding material -> attribute indirection to EXT_pbr_attributes

### PR DESCRIPTION
**Pull request by speedym**
This pull request adds the material -> attribute indirection, which improves artist workflow in DCC apps, and prevents data loss and unneeded / unexpected conversions when saving / loading .glTF files.

It allows artists to define and real-time engines to blend / switch between multiple materials, while at the same time keeping the data representation compact and optimal.

Please review it (@fhoenig, @abductee, @mmalex, @bghgary, @warrenm, @Nosferalatu, @donrmccurdy) , and if everyone agrees, I can polish it up and submit the final version.

Thanks,
Milan,
Game / VR / AR workflows lead dev.,
Modo Team, Foundry

### CODE REVIEW

**Comment by snagy** [comment]
Could you elaborate more on the benefits of the vertex attribute packing and configurable swizzling for your use cases? Adding this to the extension would definitely increase the implementation complexity for engines which wish to support it and probably lower the overall adoption rate.

And just to be clear, this is only intended to add swizzling for vertex attributes, not textures, right?

**Comment by donmccurdy** [comment]
Hi, thanks for the suggestions!

These changes make sense for DCC apps but do complicate loading. The original PR has nice properties for optimizing in WebGL engines, which are weakened if we go this direction. Retaining authoring data within a DCC pipeline is a tradeoff w/r/t efficient loading, and glTF tries to put more responsibility on the authoring tools for that reason, so I'd hope to move carefully here.

For my $0.02, it would be better to scope the changes here as a new extension that is distinct from adding metal/rough vertex attributes. An extension like `MODO_materials_pbr` sounds like the right place to start.

**Comment by speedym** [comment]
Hi Scott & Don, thank you for your review and feedback, it's highly appreciated! 

The main thing which does not make sense from the DCC / art workflow perspective in the current draft is meshes / primitives defining material properties implicitly. There should be a clear distinction between mesh / primitive data and material data. So that's where proposed indirection and matching attributes with material parameters by name comes in.

I think it's useful for cases in which artists would like to save out multiple materials and apply them to the same mesh. For example, in unlit / baked workflow, artist can create two base color maps, called "day" and "night", and simply switch between them based on the current in-game time. In the current proposal, DCC app would have to export 2 separate meshes with different COLOR_0 base color maps, and all other geometry the same, which blows up the file size. This can be quite wasteful, esp. when glTF is loaded from a remote server (say, to a web browser).

Furthermore, loading such glTF file back into DCC app would result in two meshes appearing where previously there was one with two materials, which is quite unintuitive, and it feels like a big mess to handle on DCC side, and would confuse the whole workflow.

Regarding swizzle and packing features, I thought they would be easy to implement, similar to how glTF 2.0 roughnessMetallic textures are well packed and swizzled by default. I also thought by packing them properly, less additional GL vertex attributes would be used, which could mean more performance on lower end devices.

To sum it up, I'm very much for mesh vs. material separation, and if you think swizzling & packing brings too much complexity, I could cut it out from my proposal.

**Comment by donmccurdy** [comment]
I agree mesh vs material separation is a reasonable idea, I'm just hesitant to bundle it with this particular extension. For what it's worth, there are some other options provided by the spec to consider:

1. Multiple meshes (with different materials) may contain primitives that point to the same accessors. Can't guarantee that all engines will parse this efficiently and reuse the mesh, but for example three.js makes a point of doing so. And in any case the transmitted data is not duplicated. 
  * (This is a little awkward and too implicit, but should work; perhaps there are ways of addressing it in an isolated extension or as a future spec update.)
2. Materials may be included in the file but not referenced by a mesh. This feature could be used to define a simpler extension, say, to switch materials on some activation event:

```js
"meshes": [
  {
    "name": "button_with_activation_effect",
    "primitives": [ {
      "attributes": { ... },
      "mode": 4,
      "material": 0,
      "extensions": {
        "CUSTOM_event_with_effect" {
          "event": "gaze",
          "material": 1
        }
      }
    } ],
  }
}
```

**Comment by speedym** [comment]
Coming from the DCC world, mixing materials and meshes is a big no-no, so I'm worried, because current draft goes further down that route, and messes up / complicates often used game art workflows.

I think at least having a trivial name based lookup from material property -> primitive attribute would go a long way and IMHO would not complicate real-time engine code almost at all.

**Comment by fhoenig** [comment]
Hmm. I think generalizing the attribute naming is very detrimental to the idea here. That already exists with "_" prefix. This extension is written to honor GLTF's default metal/rough model and I believe different models should go into a separate extension, similar to pbrSpecularGloss.

**Comment by speedym** [comment]
Florian, you're right, I somehow missed the _ section of the attributes spec. I'll update my proposal accordingly, and start using _ prefix to reference user defined attributes. I'll also remove the swizzle and packing features, due to Don's concerns about over-complicating implementations.

**Comment by speedym** [comment]
I have modified the extension modification proposal as per your input. Please check it out once more. (@fhoenig, @abductee, @mmalex, @bghgary, @warrenm, @nosferalatu, @donmccurdy)

**Comment by speedym** [comment]
The gist of the remaining change to original proposal is indirection from material properties -> mesh vertex attributes. If textures can be selected via materials, I don't see why not do the same for attributes.

The cost is one name based lookup, so performance wise, it should not be a concern, while it brings a lot of run-time, load & save flexibility.

**Comment by fhoenig** [comment]
I see your point. However, this is not true for COLOR_0. So which one to
stay consistent with is the open discussion here. I'm leaning towards
keeping it consistent with the way color is handled. Your proposal would
make a lot more sense if the core spec addressed the vertex color in the
same way (referenced by the material).

In my opinion, we keep it as is and defer the decision to a future core
version of GLTF, in which case the extension would then retire.

makes sense?

**Comment by speedym** [comment]
Hello Florian,

	We have a bunch of cool game art workflows which will be complicated
	by that decision. 

	If you have a bit of time to spare (and are interested), maybe we
	should meet up online, and I can screen-share and demonstrate the
	potential workflows and the complications we'll encounter if we go
	down the COLOR_0 route?

	Those workflows include instancing, replicators & multi-attribute
	/ multi-material runtime usage.

	The other way around would be to keep your extension as is, and
	file a pull request for, say, "EXT_material_vertex_attributes"
	which simply enables the indirection we need, and makes life
	easier for us.

	Thanks,
	Milan
